Allow no space in doublestar when math operation (#538) (#2095)

dpalmasan opened the pull request:
Simple fix for (#538), changes include:

- Allow no space `doublestar` in math operations
- Add tests for such scenario in which the option is provided.

The approach is simple, but it might be a good solution for the simple case of `**` as operator. When creating the destination contents, it will check on `**` based on `Line` `leaves` and recreate the line changing the `prefix` for the requested tokens (e.g. `a**b` is found, then `**` has a space prefix and `b` has a space prefix, we just re-prefix with empty `""` string). If there is just one `**` it will hug the operands.

JelleZijlstra commented on tests/data/doublestar_no_spaces.py:
```diff
@@ -0,0 +1,15 @@
+
+
+def function(**kwargs):
+    t = a**2 + b**3
```
Should also test what happens if there *is* a space in the input.

dpalmasan replied:
Maybe is it more a naming convention for the new option, maybe `skip` the chosen name is not the best, and I can change it. Regarding to the spaces question, it will yield the same output (meaning, `DOUBLESTAR` without spaces).

Does your comment mean, update the code and not merge yet, or does it mean: These changes are not convenient, forget about them?

I am sorry if I don't understand correctly it is my first PR on an open-source repo.

felix-hilden replied:
Not sure, but I think Jelle meant you should add test cases which have spaces around the operators, which I think you've done in the code just below! @JelleZijlstra what do you reckon?

I'm not sure about the test writing practices of Black, but I see lots of tests being written a bit more condensed. In your file it's nice to see immediately what's expected from the function name, but another option would be to simply have all the cases in one succession. That's up for debate though!

JelleZijlstra reviewed (commented):
The new option is actually a bit different than existing `--skip` options: it means "format things differently", not "leave this area of formatting alone", as the previous options did.

I'd prefer we don't merge this change.

ichard26 commented:
To second Jelle's dislike, I would not be happy merging this as long as the new flag is introduced. Black is growing a lot of new options and this makes the situation worse. The lack of configuration is a defining feature of Black and I wouldn't be pleased to see that wither away. Because I'm quite busy ATM, I recommend reading the discussion on this PR: https://github.com/psf/black/pull/1904 where in particular I share the same concerns: https://github.com/psf/black/pull/1904#issuecomment-757101552.

Thanks for the PR and the time investment.

dpalmasan commented:
Okay, I understand. The only reason I added the flag, was to not alter the default behavior. However, if I remove the flag and always do that on `DOUBLESTAR` operators, would that suffice for a #538 fix?

Thanks for the reviews!

dpalmasan commented:
Okay, I updated the PR removing the flag. I also added what was discussed on (#538 ) if just one operator is found, not hug the operands with the operators.

felix-hilden commented:
As per #53, I'm asking the maintainers: should the changes here be also added to the Black code style document?

ichard26 commented on src/black/__init__.py:
```diff
@@ -1078,6 +1078,7 @@ def f(
         for line in transform_line(
             current_line, mode=mode, features=split_line_features
         ):
+            line = fix_doublestar_in_op(line)
```
I'm a little concerned by the performance impact this will have. Doing some rough testing, formatting `src/black/__init__.py` spends ~6500ms in the `format_str` method and ~4% (~300ms) of the total runtime was spent in the `fix_doublestar_in_op`... which is IMO too much for such a simple case like this. 

<details>

<summary>Profiling data (warning: extremely large image)</summary>

![out](https://user-images.githubusercontent.com/63936253/114799184-08cf6780-9d65-11eb-8c1a-7b437e1ee83d.png)

</details>

Is it possible to make this more efficient? Cloning the leaf seems to be where most of 300ms cost is coming from. Perhaps moving the logic somewhere into one of the transformers called in `transform_line` could help since some lines might not need this fixing? I'm sorry for my unadvice, I usually don't review formatting code.

dpalmasan replied:
Ohh okay, gonna take a look. Thanks for pointing out!

ambv replied:
Agreed with @ichard26, it would be cleaner if it were a StringTransformer like the other transformation. This would additionally solve the performance issue. Few enough lines in Python have DOUBLESTAR tokens that the simplest way to improve performance is just look whether there even is doublestar on the line in `do_match()`.

ichard26 reviewed (commented):
While I can't review the actual formatting code since I'm too inexperienced, I do have other comments I'd like to mention. 

Thank you for the time and effort!

felix-hilden commented on CHANGES.md:
```diff
@@ -4,6 +4,8 @@
 
 #### _Black_
 
+- `Black` wraps `**` if in math expression with more than 1 op (#2095)
```
I like that more verbose wording too! And since it's not updated yet let's grammar-nazi it up just a wee bit: "unless it's *the* only operation". First I thought it read something like "unless it's *just* operation" which confused me a bit. Obvious on closer inspection but worth a correction in my mind 😄

felix-hilden replied:
Damn this should have been a reply to @ichard26, see his suggested changes above.

ambv commented on tests/data/expression.diff:
```diff
@@ -19,7 +19,7 @@
 -+really ** -confusing ** ~operator ** -precedence
 -flags & ~ select.EPOLLIN and waiters.write_task is not None
-++(really ** -(confusing ** ~(operator ** -precedence)))
+++(really**-(confusing**~(operator**-precedence)))
```
Wait... what? You really think this is more readable in the new form? I think that once you hit a complex base or exponent, then there is no value in skipping the spaces.

tupui commented:
Hey what is the status here?

As Ralf noted in the linked issue,`**`is not covered by the PEP8 and most (all) codes like NumPy/SciPy never add space around this operator.

In general there it would be great to allow more flexibility around for writing maths. Nobody writes maths like black does... Also remember the second section of PEP8 https://www.python.org/dev/peps/pep-0008/#a-foolish-consistency-is-the-hobgoblin-of-little-minds

ichard26 commented:
Congrats @dpalmasan your change has found its way into Black eventually :tada: Thanks for submitting this PR!

Yikun commented:
Sorry to post comments, just curious, I would like to know have we discussed the case of `x ** 'a string'`, looks like it still not allow in current version, looks a bit inconsistent?

```
# Not allow
3 ** "a string"
# Allow
2**63 - 1
```

ichard26 commented:
@Yikun I mean, that seems like such an edge case since strings can't be a valid operand for power operations with any of the built-in types AFAIK. Do you have a real life example where you'd write this kind of code?
